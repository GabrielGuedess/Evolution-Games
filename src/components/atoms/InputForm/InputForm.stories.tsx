import { Story, Meta } from '@storybook/react/types-6-0';
import { Envelope } from 'phosphor-react';

import { InputForm, InputFormProps } from './InputForm';

export default {
  title: 'Atoms/InputForm',
  component: InputForm,
  args: {
    isInvalid: false,
    errorMessage: 'Campo obrigatório',
  },
} as Meta<InputFormProps>;

export const Default: Story<InputFormProps> = args => (
  <div style={{ maxWidth: '440px' }}>
    <InputForm label="E-mail" placeholder="Insira seu email" {...args} />
  </div>
);

export const Invalid: Story<InputFormProps> = args => (
  <div style={{ maxWidth: '440px' }}>
    <InputForm label="E-mail" placeholder="Insira seu email" {...args} />
  </div>
);

Invalid.args = {
  isInvalid: true,
  errorMessage: 'Campo obrigatório',
};

export const WithIcon: Story<InputFormProps> = args => (
  <div style={{ maxWidth: '440px' }}>
    <InputForm label="Senha" placeholder="Insira sua senha" {...args} />
  </div>
);

WithIcon.args = {
  icon: <Envelope size={28} />,
};
