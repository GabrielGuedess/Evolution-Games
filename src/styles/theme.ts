export default {
  grid: {
    container: '165.6rem',
  },
  border: {
    radius: {
      regular: '1rem',
    },
  },
  font: {
    family: {
      primary:
        "'Inter', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
    },
    regular: 400,
    medium: 500,
    sizes: {
      xxsmall: '1rem',
      xsmall: '1.2rem',
      small: '1.4rem',
      medium: '1.6rem',
      xmedium: '1.8rem',
      xxmedium: '2.0rem',
      big: '2.2rem',
      xbig: '2.5rem',
      xxbig: '2.8rem',
      xxxbig: '3.0rem',
      large: '3.2rem',
      xlarge: '4.5rem',
      xxlarge: '5.0rem',
      extra: '6.0rem',
    },
  },
  colors: {
    primary: '#7C3AED',
    secondary: '#717B8A',
    mainBg: '#04040C',
    white: '#FFFFFF',
    whiteText: '#F4F4F5',
    gameDetails: '#010914',
    gameGreat: '#0FBF00',
    gameAverage: '#FFF720',
    gameBad: '#FF3333',
  },
  spacings: {
    xxsmall: '0.8rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem',
    big: '5.8rem',
    huge: '8.0rem',
    section: '10rem',
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50,
  },
  shadows: {
    text: '0 0.8rem 1rem rgba(118, 74, 241, 0.5)',
    box: '0 0.8rem 2.8rem rgba(118, 74, 241, 0.5)',
  },
} as const;
