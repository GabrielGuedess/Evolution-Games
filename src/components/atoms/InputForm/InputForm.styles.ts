import media from 'styled-media-query';

import styled, { css } from 'styled-components';

import { InputFormProps } from './InputForm';

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

export const WrapperInput = styled.div`
  position: relative;
`;

export const Label = styled.label`
  ${({ theme }) => css`
    position: absolute;
    left: 1rem;
    top: 1.7rem;
    padding: 0 3px;
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.regular};
    transition: all 0.3s;
    z-index: 2;

    ::before {
      content: '';
      position: absolute;
      inset: 0;
      width: 100%;
      z-index: -1;

      ${media.greaterThan('medium')`
        background: ${theme.colors.mainBg};
      `}
    }
  `}
`;

export const Input = styled.input<Pick<InputFormProps, 'isInvalid'>>`
  ${({ theme, isInvalid }) => css`
    width: 100%;
    font-size: ${theme.font.sizes.xmedium};
    font-weight: ${theme.font.regular};
    font-style: normal;
    line-height: ${theme.spacings.medium};
    border-width: 2px;
    border-style: solid;
    border-color: ${isInvalid ? theme.colors.gameBad : '#272530'};
    background-color: transparent;
    border-radius: 0.5rem;
    padding: ${theme.spacings.xxsmall};
    color: ${theme.colors.white};
    outline: none;
    transition: all 0.3s;

    ::-webkit-calendar-picker-indicator {
      background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="16" height="15" viewBox="0 0 24 24"><path fill="%23272530" d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"/></svg>');
      transition: all 0.3s;
    }

    ${Label} {
      color: ${isInvalid ? theme.colors.gameBad : theme.colors.white};
    }

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0px 100rem ${theme.colors.mainBg} inset;
      -webkit-text-fill-color: ${theme.colors.white};
    }

    &:-webkit-autofill ~ div,
    &:-webkit-autofill:hover ~ div,
    &:-webkit-autofill:focus ~ div,
    &:-webkit-autofill:active ~ div {
      color: ${theme.colors.secondary};
    }

    &::placeholder {
      font-size: ${theme.font.sizes.small};
      opacity: 0;
      transition: all 0.3s;
    }

    &:focus {
      border-color: ${isInvalid ? theme.colors.gameBad : theme.colors.primary};

      ::-webkit-calendar-picker-indicator {
        background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="16" height="15" viewBox="0 0 24 24"><path fill="%237C3AED" d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"/></svg>');
      }

      ::placeholder {
        opacity: 1;
      }
    }

    &:is(:focus, :valid) + ${Label} {
      top: -2rem;
      color: ${isInvalid ? theme.colors.gameBad : theme.colors.primary};
      font-size: ${theme.font.sizes.xsmall};

      ${media.greaterThan('medium')`
        top: -8px;
      `}
    }

    ${isInvalid &&
    css`
      &:invalid + ${Label} {
        top: -2rem;
        color: ${theme.colors.gameBad};
        font-size: ${theme.font.sizes.xsmall};

        ${media.greaterThan('medium')`
          top: -8px;
        `}
      }
    `}

    ::selection {
      color: ${theme.colors.white};
      background: ${theme.colors.primary};
    }
  `}
`;

export const IconWrapper = styled.div`
  width: 50px;
  height: 50px;
  display: flex;
  align-items: center;
  justify-content: center;
  position: absolute;
  right: 0;
  top: 0;
`;

export const ErrorMessage = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    margin-left: 0.2rem;
    color: ${theme.colors.gameBad};

    > span {
      font-size: ${theme.font.sizes.xsmall};
      font-style: normal;
      font-weight: ${theme.font.regular};
      margin: ${theme.spacings.xxsmall} 0.4rem;
    }
  `}
`;
