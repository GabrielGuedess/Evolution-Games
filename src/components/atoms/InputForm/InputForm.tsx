import { InputHTMLAttributes } from 'react';

import { WarningCircle } from 'phosphor-react';

import * as S from './InputForm.styles';

export type InputFormProps = {
  label?: string;
  labelFor?: string;
  icon?: JSX.Element;
  isInvalid?: boolean;
  errorMessage?: string;
  mask?: string;
  as?: React.ElementType;
} & InputHTMLAttributes<HTMLInputElement>;

export const InputForm = ({
  label,
  isInvalid = false,
  labelFor,
  icon,
  errorMessage,
  ...props
}: InputFormProps) => (
  <S.Wrapper>
    <S.WrapperInput>
      <S.Input
        autoComplete="on"
        name={labelFor}
        id={labelFor}
        isInvalid={isInvalid}
        {...props}
      />
      {!!label && (
        <S.Label htmlFor={labelFor} aria-label={labelFor}>
          {label}
        </S.Label>
      )}

      {!!icon && <S.IconWrapper>{icon}</S.IconWrapper>}
    </S.WrapperInput>

    {isInvalid && !!errorMessage && (
      <S.ErrorMessage>
        <WarningCircle size={18} />
        <span>{errorMessage}</span>
      </S.ErrorMessage>
    )}
  </S.Wrapper>
);
