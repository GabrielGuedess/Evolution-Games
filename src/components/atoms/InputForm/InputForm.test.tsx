import { Envelope } from 'phosphor-react';
import { renderWithTheme } from 'utils/tests/helpers';

import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import theme from 'styles/theme';

import { InputForm } from './InputForm';

describe('<InputForm />', () => {
  it('should render the <InputForm />', () => {
    // Arrange
    const { container } = renderWithTheme(<InputForm label="Senha" />);

    // Assert
    expect(screen.getByText('Senha')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the <InputForm /> with the indication of invalid values', () => {
    renderWithTheme(
      <InputForm
        isInvalid
        placeholder="E-mail"
        label="E-mail"
        labelFor="email"
        icon={<Envelope size={20} />}
        errorMessage="Campo obrigatório"
      />,
    );

    // Assert
    expect(screen.getByLabelText('email')).toHaveStyle({
      color: theme.colors.gameBad,
    });

    expect(screen.getByText('Campo obrigatório')).toBeInTheDocument();
    expect(screen.getByText('E-mail')).toBeInTheDocument();
  });

  it('should change colors with invalid email', async () => {
    renderWithTheme(
      <InputForm
        isInvalid
        placeholder="E-mail"
        label="E-mail"
        labelFor="email"
        icon={<Envelope size={20} />}
        errorMessage="Campo obrigatório"
      />,
    );

    // Arrange
    const inputEmail = screen.getByPlaceholderText(
      'E-mail',
    ) as HTMLInputElement;

    // Act
    await userEvent.type(inputEmail, 'test');

    // Assert
    waitFor(() => {
      expect(screen.getByLabelText('email')).toHaveStyle({
        color: theme.colors.primary,
      });
    });
  });
});
