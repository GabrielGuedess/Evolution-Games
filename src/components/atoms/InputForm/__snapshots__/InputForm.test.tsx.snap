// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['<InputForm /> should render the <InputForm /> 1'] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  position: relative;
}

.c4 {
  position: absolute;
  left: 1rem;
  top: 1.7rem;
  padding: 0 3px;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  z-index: 2;
}

.c4::before {
  content: '';
  position: absolute;
  inset: 0;
  width: 100%;
  z-index: -1;
}

.c2 {
  width: 100%;
  font-size: 1.8rem;
  font-weight: 400;
  font-style: normal;
  line-height: 3.2rem;
  border-width: 2px;
  border-style: solid;
  border-color: #272530;
  background-color: transparent;
  border-radius: 0.5rem;
  padding: 0.8rem;
  color: #FFFFFF;
  outline: none;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c2::-webkit-calendar-picker-indicator {
  background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="16" height="15" viewBox="0 0 24 24"><path fill="%23272530" d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"/></svg>');
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c2 .c3 {
  color: #FFFFFF;
}

.c2:-webkit-autofill {
  -webkit-box-shadow: 0 0 0px 100rem #04040C inset;
  -webkit-text-fill-color: #FFFFFF;
}

.c2:-webkit-autofill ~ div,
.c2:-webkit-autofill:hover ~ div,
.c2:-webkit-autofill:focus ~ div,
.c2:-webkit-autofill:active ~ div {
  color: #717B8A;
}

.c2::-webkit-input-placeholder {
  font-size: 1.4rem;
  opacity: 0;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c2::-moz-placeholder {
  font-size: 1.4rem;
  opacity: 0;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c2:-ms-input-placeholder {
  font-size: 1.4rem;
  opacity: 0;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c2::placeholder {
  font-size: 1.4rem;
  opacity: 0;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c2:focus {
  border-color: #7C3AED;
}

.c2:focus::-webkit-calendar-picker-indicator {
  background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="16" height="15" viewBox="0 0 24 24"><path fill="%237C3AED" d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"/></svg>');
}

.c2:focus::-webkit-input-placeholder {
  opacity: 1;
}

.c2:focus::-moz-placeholder {
  opacity: 1;
}

.c2:focus:-ms-input-placeholder {
  opacity: 1;
}

.c2:focus::placeholder {
  opacity: 1;
}

.c2:is(:focus,:valid) + .c3 {
  top: -2rem;
  color: #7C3AED;
  font-size: 1.2rem;
}

.c2::selection {
  color: #FFFFFF;
  background: #7C3AED;
}

@media (min-width:768px) {
  .c4::before {
    background: #04040C;
  }
}

@media (min-width:768px) {
  .c2:is(:focus,:valid) + .c3 {
    top: -8px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      autocomplete="on"
      class="c2"
    />
    <label
      class="c3 c4"
    >
      Senha
    </label>
  </div>
</div>
`;
